apps/api_web/lib/api_web/event_stream/canary.ex

```diff
@@ -0,0 +1,26 @@
+defmodule ApiWeb.EventStream.Canary do
+  @moduledoc """
+  Process intended to be placed after the Endpoint in the same supervision tree. When terminated
+  due to a shutdown, it notifies EventStream processes, allowing them to cleanly close their HTTP
+  connections before they are themselves terminated by the subsequent shutdown of the Endpoint.
+  """
+
+  use GenServer
+
+  @default_notify_fn &ApiWeb.EventStream.Supervisor.subscriber_dispatch/1
+
+  def start_link(notify_fn \\ @default_notify_fn) do
+    GenServer.start_link(__MODULE__, notify_fn, [])
+  end
+
+  @impl true
+  def init(notify_fn) do
+    Process.flag(:trap_exit, true)
+    {:ok, notify_fn}
+  end
+
+  @impl true
+  def terminate(:shutdown, notify_fn), do: notify_fn.(:shutdown)
+  def terminate({:shutdown, _reason}, notify_fn), do: notify_fn.(:shutdown)
```

Comment:
since you only call `notify_fn/1` with `:shutdown` does it need to take an argument?

Reply:
I guess it's sort of a separation-of-concerns thing. The normal value for this function is `subscriber_dispatch`, which just forwards the provided message to all subscribers, not caring about what it is; the `Supervisor` doesn't need to know it's specifically being used to message a shutdown. I might instead call this variable `dispatch_fn` to make it clear it's expected to be a generic thing.